### PR TITLE
Add callbacks

### PR DESCRIPTION
**Pull request by brettshollenberger**
Hey Andrew! Thanks for this library 😄 

I wanted to add the [callbacks API](https://xgboost.readthedocs.io/en/stable/python/python_api.html#callback-api) so I could support an integration with [Wandb](https://github.com/wandb/wandb/blob/48dbe87faca6ad1f9fbfe7dd15c2a22e651cf145/wandb/integration/xgboost/xgboost.py) 

I have the Ruby implementation of this [callback over here](https://github.com/brettshollenberger/wandb_rb/blob/main/lib/wandb/xgboost_callback.rb) as an example of the use case and have tested the integration in my Wandb console

Some dummy data just to show the integration:
Screenshot 2024-10-11 at 2 36 17 PM
Screenshot 2024-10-11 at 2 35 55 PM

I kept the API consistent w/ the Python implementation but let me know if there's anything else you'd want to see here

### CODE REVIEW

**Comment by ankane** [comment]
Hey Brett, great to hear from you! This looks pretty neat. It looks like the Python API uses the return values of the callbacks in `train`. I think the best way to do that would be to add `TrainingCallback` and `CallbackContainer` classes like Python ([code](https://github.com/dmlc/xgboost/blob/v2.1.1/python-package/xgboost/callback.py)), but I can do that in a follow-up commit if you just want to get the tests passing.

**Comment by brettshollenberger** [comment]
Good idea! Let me know what you think, should be passing now

File: lib/xgboost.rb [hunk 1]
```diff
@@ -44,8 +44,12 @@ class << self
   autoload :FFI, "xgboost/ffi"
 
   class << self
-    def train(params, dtrain, num_boost_round: 10, evals: nil, early_stopping_rounds: nil, verbose_eval: true)
+    def train(params, dtrain, num_boost_round: 10, evals: nil, early_stopping_rounds: nil, verbose_eval: true,
+              callbacks: [])
       booster = Booster.new(params: params)
+      cb_container = CallbackContainer.new(callbacks)
+      cb_container.before_training(model: booster)
```
Comment (ankane) [inline lib/xgboost.rb:51]: This should update `booster` to match Python.

File: lib/xgboost.rb [hunk 2]
```diff
@@ -59,32 +63,36 @@ def train(params, dtrain, num_boost_round: 10, evals: nil, early_stopping_rounds
       end
 
       num_boost_round.times do |iteration|
+        cb_container.before_iteration(model: booster, epoch: iteration)
```
Comment (ankane) [inline lib/xgboost.rb:66]: This should break if the return value is falsy.

File: lib/xgboost.rb [hunk 3]
```diff
@@ -59,32 +63,36 @@ def train(params, dtrain, num_boost_round: 10, evals: nil, early_stopping_rounds
...
         booster.update(dtrain, iteration)
 
-        if evals.any?
```
Comment (ankane) [inline lib/xgboost.rb:64]: Please keep the existing code where possible to keep the changeset minimal / easier to review.

File: lib/xgboost.rb [hunk 4]
```diff
@@ -59,32 +63,36 @@ def train(params, dtrain, num_boost_round: 10, evals: nil, early_stopping_rounds
...
+          best_score = score
+          best_iter = iteration
+          best_message = message
+        elsif early_stopping_rounds && iteration - best_iter >= early_stopping_rounds
+          booster.best_iteration = best_iter
+          booster.best_score = best_score
+          puts "Stopping. Best iteration:\n#{best_message}" if verbose_eval
+          break
         end
+        cb_container.after_iteration(model: booster, epoch: iteration, res: res)
```
Comment (ankane) [inline lib/xgboost.rb:93]: This should `break` for falsy values like `before_iteration`.

File: lib/xgboost.rb [hunk 5]
```diff
@@ -59,32 +63,36 @@ def train(params, dtrain, num_boost_round: 10, evals: nil, early_stopping_rounds
...
       end
+      cb_container.after_training(model: booster)
```
Comment (ankane) [inline lib/xgboost.rb:95]: This should update `booster` like `before_training`.

File: lib/xgboost/callback_container.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
+module XGBoost
+  class CallbackContainer
+    attr_reader :callbacks, :history
+
+    def initialize(callbacks)
+      @callbacks = callbacks
+      @history = {}
+
+      callbacks.each do |callback|
+        unless callback.class.ancestors.include?(TrainingCallback)
```
Comment (ankane) [inline lib/xgboost/callback_container.rb:10]: `callback.is_a?(TrainingCallback)`

File: test/train_test.rb [hunk 7]
```diff
@@ -55,6 +55,88 @@ def test_feature_names_and_types
     assert_nil model.feature_types
   end
 
+  class MockCallback < XGBoost::TrainingCallback
```
Comment (ankane) [inline test/train_test.rb:58]: Let's create a separate file for the callback tests.

**Review by ankane — COMMENTED** [review]
Awesome, thanks for adding those classes. Added some comments inline. It looks like the new files need to be required.

```ruby
require_relative "xgboost/callback_container"
require_relative "xgboost/training_callback"
```

Also, if you're seeing an error running the tests locally, try running:

```sh
bundle exec rake vendor:platform
```

**Comment by brettshollenberger** [comment]
Thanks Andrew, had to get rid of some aggressive Rubocop settings complicating the changes. Tests are all passing for me and everything's committed now 😅 Looks like you have to approve the Github workflow run here but I think we should have it this time.

File: lib/xgboost.rb (hunk 3 above)
Comment (brettshollenberger) [reply lib/xgboost.rb:64]: Sorry, disabled rubocop

**Comment by ankane** [comment]
Thanks @brettshollenberger! Merged in the commit above with a few minor changes:

1. I had the logic backwards for `before`/`after_iteration` - it should stop if it returns a truthy value
2. Changed callbacks to use positional arguments instead of keyword
3. Left out the `params` change, as it's not present in the Python library (from what I can tell)

Going to spend a little time getting the overall code more in sync with Python, and then will push a new release.

**Comment by brettshollenberger** [comment]
@ankane nice, thanks! 

With the params thing, it's not exactly 1:1, but I think an easier solution than the Python Wandb callback, which calls `xgboost/core#save_config`. 

https://github.com/wandb/wandb/blob/8698af5862e44baf31af5411b81bea546e069257/wandb/integration/xgboost/xgboost.py#L117

https://github.com/dmlc/xgboost/blob/3f9bfaf86e6db6a4f54734aa7d164df55aa69ef6/python-package/xgboost/core.py#L2008

```python
    def before_training(self, model: Booster) -> Booster:
        """Run before training is finished."""
        # Update W&B config
        config = model.save_config()
        wandb.config.update(json.loads(config))

        return model
```

```python
    def save_config(self) -> str:
        """Output internal parameter configuration of Booster as a JSON
        string.

        .. versionadded:: 1.0.0

        """
        json_string = ctypes.c_char_p()
        length = c_bst_ulong()
        _check_call(
            _LIB.XGBoosterSaveJsonConfig(
                self.handle, ctypes.byref(length), ctypes.byref(json_string)
            )
        )
        assert json_string.value is not None
        result = json_string.value.decode()  # pylint: disable=no-member
        return result
```

**Comment by ankane** [comment]
I think it's better to keep things synced for maintainability in most cases. Added `save_config` in the commit above.

**Comment by brettshollenberger** [comment]
Awesome, thank you!

**Comment by ankane** [comment]
Great, just pushed 0.9.0. Let me know if you need anything else that's missing.

**Pull request closed without merging**